src/Slick/Build.hs

```diff
@@ -0,0 +1,103 @@
+{-# LANGUAGE DeriveAnyClass    #-}
+{-# LANGUAGE DeriveGeneric     #-}
+{-# LANGUAGE OverloadedStrings #-}
+
+module Slick.Build
+  ( convert
+  , srcToURL
+  , destToSrc
+  , srcToDest
+  , flattenMeta
+  , pruner
+  , EntityFilePath(..)
+  , shakeArgsAlwaysPruneWith
+  ) where
+
+import           Control.Lens
+import           Control.Monad
+import           Data.Aeson
+import           Data.Aeson.Lens
+import           Data.List                  as L (intersperse, sortBy, (\\))
+import           Data.Maybe
+import           Data.Monoid
+import qualified Data.Text                  as T
+import           Development.Shake
+import           Development.Shake.Classes
+import           Development.Shake.FilePath
+import           GHC.Generics               hiding (Meta)
+import           System.Console.GetOpt
+import           System.Directory.Extra     (listFilesRecursive, removeFile)
+import           System.IO.Extra            as IO
+import           Text.Pandoc
+import           Text.Pandoc.Highlighting
+import           Text.Pandoc.Shared
+
+import           Slick.Utils
+
+--------------------------------------------------------------------------------
+
+-- | Generalized version of filepath
+newtype EntityFilePath a =
+  EntityFilePath String
+   deriving (Show, Eq,  Generic, Hashable, Binary, NFData)
+
+-- | convert 'build' filepaths into source file filepaths
+destToSrc :: FilePath -> FilePath
+destToSrc p = "site" </> dropDirectory1 p
+
+-- | convert source filepaths into build filepaths
+srcToDest :: FilePath -> FilePath
+srcToDest p = "dist" </> dropDirectory1 p
+
+-- | convert a source file path into a URL
+srcToURL :: FilePath -> String
+srcToURL = ("/" ++) . dropDirectory1 . (-<.> ".html")
+
+--------------------------------------------------------------------------------
+
+-- | Attempt to convert between two JSON serializable objects (or 'Value's).
+--   Failure to deserialize fails the Shake build.
+convert :: (FromJSON a, ToJSON a, FromJSON b) => a -> Action b
+convert a = case fromJSON (toJSON a) of
+  Success r   -> pure r
+  Error   err -> fail $ "json conversion error:" ++ err
+
+-- | Flatten a Pandoc 'Meta' into a well-structured JSON object, rendering Pandoc
+--   text objects into plain strings along the way.
+flattenMeta :: Meta -> Value
+flattenMeta (Meta meta) = toJSON $ fmap go meta
+ where
+  go :: MetaValue -> Value
+  go (MetaMap     m) = toJSON $ fmap go m
+  go (MetaList    m) = toJSONList $ fmap go m
+  go (MetaBool    m) = toJSON m
+  go (MetaString  m) = toJSON m
+  go (MetaInlines m) = toJSON $ stringify m
+  go (MetaBlocks  m) = toJSON $ stringify m
+
+--------------------------------------------------------------------------------
+
+-- | A version of 'shakeArgsPrunWith' that always do prune at the end.
```

Comment:
```suggestion
-- | A version of 'shakeArgsPruneWith' that always does pruning at the end.
```

Please mention how the pruner works, or at the least refer to `'pruner'` in the description.

Have you had need to customize the pruner in your own stuff? I'd prefer to keep the number of options and customizations simple, exposing additional helpers where needed, it really helps with adoption. Most folks won't be diving in as deeply as you 😉, in this case we should inline `pruner` and not even expose it if we can.